app/scripts/platforms/extension.js

```diff
@@ -65,9 +65,11 @@ class ExtensionPlatform {
   }
 
   sendMessage (message, query = {}) {
-    extension.tabs.query(query, tabs => {
+    const id = query.id
+    delete query.id
+    extension.tabs.query({ ...query }, tabs => {
       tabs.forEach(tab => {
-        extension.tabs.sendMessage(tab.id, message)
+        extension.tabs.sendMessage(id || tab.id, message)
```

Comment:
If the `id` is passed in and we're querying all the tabs, will this send the same message to the same tab multiple times?

Reply:
Should we be sending a message to tab `id` if that's passed in, else iterate all the tabs?

Reply:
I feel like a query with multiple properties should hit tabs that satisfy _all_ properties, not just some. This is why we 1) query without the ID, then 2) iterate through the results and further filter by ID. If we detected ID first then bailed, that implies an ID takes precedence over all other query properties.

Reply:
Hmm, maybe I don't understand what these APIs are doing well enough. If we call `platform.sendMessage(foo, 42)` we'll be calling `extension.tabs.sendMessage(42, foo)` multiple times (once for each tab returned in the query, no?

Reply:
Whymarrh's concern made sense to me, and then I tried to word it and understood Paul's point:

The [tabs.query method](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/tabs/query) does not take an `id` parameter and filter for it, and so this basically allows our own `platform.sendMessage()` method to accept an `id` in its query, and then only send to that tab.